Switch to “intents” API

katzmo opened the pull request.

katzmo committed:
Switching to new Intents API: basic one-off donation flow

katzmo committed:
Refactor JS: migrate to Parcel

katzmo committed:
Split card form over multiple fields

Use dedicated Stripe elements (card number etc)
instead of a single "card" field for better UX.

katzmo committed:
Use custom field type for Stripe fields

katzmo committed:
Pass actual input styles to Stripe elements.

katzmo committed:
Update additional Stripe fields

* Update existing address field names
* Add email and phone fields
* Send billing data to Stripe
* Fix data name issue

katzmo committed:
Register webhooks for payment methods on Stripe.

katzmo committed:
Implement webhook callback to update payment status.

Stripe expects a timely answer with a 2xx response code for successful
webhook calls. The response body is irrelevant but visible in the
Stripe Dashboard.

katzmo committed:
Add custom payment status "accepted".

Changed to success (or failed) when the payment has been processed by
Stripe.

katzmo committed:
Make sure to always load library before using it.

katzmo committed:
Use payment_recurrence for recurring payments.

torotil commented on src/CreditCardForm.php:
```diff
@@ -35,21 +35,21 @@ public function form(array $form, array &$form_state, \Payment $payment) {
     );
 
     drupal_add_js($settings, 'setting');
+    drupal_add_js('https://js.stripe.com/v3/', 'external');
```
In general I’d prefer using #attached over direct use of `drupal_add_js()`.

torotil reviewed (commented):
… and many `phpcs` issues.

torotil committed:
Use #attached to attach scripts.

katzmo committed:
Update flow for recurring payments.

torotil committed:
Use entity-hooks to store extra payment data.

torotil committed:
Fix syntax error in CreditCardController::createCustomer().

torotil committed:
phpcs fixes.

torotil commented on src/CreditCardController.php:
```diff
@@ -45,141 +68,182 @@ function validate(\Payment $payment, \PaymentMethod $method, $strict) {
+      return SetupIntent::create();
+    }
+    // TODO: What now? Exception for "nothing to pay for"?
+  }
```
This means the payment method can’t deal with mixed line items, right? It always uses a single intent for the whole execution.

katzmo replied:
No, it’s fine to use a PaymentIntent for subscriptions too.

torotil replied:
Ok, after discussion: PaymentIntents can (and should) be used whenever there is at least one non-recurring line-item.

torotil commented on src/CreditCardController.php:
```diff
@@ -45,141 +68,182 @@ function validate(\Payment $payment, \PaymentMethod $method, $strict) {
+    // PaymentIntent: Make a payment immediately.
+    if ($this->filterRecurringLineItems($payment, FALSE)) {
+      return PaymentIntent::create([
         'amount'   => $this->getTotalAmount($payment),
```
This gets the total of all line items. Even if some of them are recurring.

katzmo replied:
That’s right, `getTotalAmout()` should only sum up one-off line-items!

torotil committed:
Refactoring: Move most of the stripe-API calling code into a Api-class.

torotil committed:
Introduce helper function to split payments into recurring / one-off.

torotil committed:
Fix string templates for plan creation.

torotil committed:
Copy approach for data fields from braintree_payment.

torotil committed:
Fix reuse of update number.

katzmo committed:
Refactor subscription creation.

torotil committed:
Fix call to Payment::totalAmount().

torotil committed:
Set amount in cent.

torotil committed:
Adapt JS to new field mapping.

torotil committed:
Recompile JS.

katzmo committed:
Update to stripe-php v7.0.

katzmo committed:
Fix stripe_payment table records.

Since the stripe_id must be unique and not null, we have to wait for an
actual stripe id before saving any data to the table.

katzmo committed:
Various fixes.

katzmo committed:
phpcs fixes and improved doc strings.

katzmo committed:
Save subscriptions to DB instead of plans.

katzmo committed:
Use library constants for webhook event names.

katzmo committed:
Fix js path

torotil committed:
Let payment UI save first before registering the webhook.

katzmo committed:
Fix subscription + plan + product creation.

katzmo committed:
Use more appropriate watchdog error levels.

katzmo committed:
Compile JS.

torotil committed:
Use a consistent API-version.

katzmo committed:
Check for a valid amount before making Stripe payment available.

torotil committed:
Set the stripe app info.

katzmo committed:
Better error handling for recurrent payments.

* Do not create incomplete payments that wait for user interaction
* Set payment status to failed when subscription creation fails
* Make sure the form is still useable after a failed payment attempt

katzmo committed:
Fix: remove leftover config from validation

katzmo committed:
Compile JS.

katzmo committed:
Improve handling of webhooks.

Stripe does not know about our payment methods and will send events for
any payments to all webhooks of this account.

* Delete webhooks as well when the payment method is deleted.
* Change payment status only once ← only the original method
  should react to webhook events.
